core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java

```diff
@@ -3301,6 +3301,85 @@ private void checkHavingAliasSameAsColumn(boolean upperAlias) {
     sql(query).ok(expected);
   }
 
+  @Test void testSetOpLimitShouldRetainParentheses() {
+    // If the subquery parentheses of limit are discarded,
+    // the semantics of parsing will be affected.
+    // Parentheses should be retained, otherwise not required.
+    final String query = "select \"product_id\" from \"product\""
+        + "union all\n"
+        + "(select \"product_id\" from \"product\" limit 10)"
+        + "intersect all\n"
+        + "(select \"product_id\" from \"product\")";
+    final String expected = "SELECT \"product_id\"\n"
+        + "FROM \"foodmart\".\"product\"\n"
+        + "UNION ALL\n"
+        + "SELECT *\n"
+        + "FROM ((SELECT \"product_id\"\n"
+        + "FROM \"foodmart\".\"product\"\n"
+        + "FETCH NEXT 10 ROWS ONLY)\n"
+        + "INTERSECT ALL\n"
+        + "SELECT \"product_id\"\n"
+        + "FROM \"foodmart\".\"product\")";
+    sql(query).ok(expected);
+
+    // Offset is the same as limit.
+    final String unionOffsetQuery = "select \"product_id\" from(\n"
+        + "(select \"product_id\"  from \"product\" limit 10)\n"
+        + "union all\n"
+        + "(select \"product_id\" from \"product\" offset 5)\n"
+        + "union all\n"
+        + "(select \"product_id\" from \"product\" order by \"product_id\" limit 5 offset 5))";
+    final String unionOffsetRes = "SELECT *\n"
+        + "FROM (SELECT *\n"
+        + "FROM ((SELECT \"product_id\"\n"
+        + "FROM \"foodmart\".\"product\"\n"
+        + "FETCH NEXT 10 ROWS ONLY)\n"
+        + "UNION ALL\n"
+        + "(SELECT \"product_id\"\n"
+        + "FROM \"foodmart\".\"product\"\n"
+        + "OFFSET 5 ROWS))\n"
+        + "UNION ALL\n"
+        + "(SELECT \"product_id\"\n"
+        + "FROM \"foodmart\".\"product\"\n"
+        + "ORDER BY \"product_id\"\n"
+        + "OFFSET 5 ROWS\n"
+        + "FETCH NEXT 5 ROWS ONLY)) AS \"t6\"";
+    sql(unionOffsetQuery).ok(unionOffsetRes);
+  }
+
+  @Test void testSetOpOrderMaybeRetainParentheses() {
+    // Order by without limit/offset will be removed, so parentheses are not required.
```

Comment:
this test is testing RelToSql, not SqlToRel. So I would remove this case (the first half of this method).

Reply:
Ok, I've deleted this test case.